mindmeld/text_preparation/text_preparation_pipeline.py

```diff
@@ -0,0 +1,530 @@
+            entity_text = match.group(1)
+
+            # Adds " {"
+            modified_text.append(text[entity_start:entity_text_start])
+
+            # Adds function(entity_text)
+            modified_text.append(function(entity_text))
+
+            # Adds "|"
+            modified_text.append(text[entity_text_end:entity_end])
+
+            # Update the previous entity ending index
+            prev_entity_end = entity_end
+
+        if prev_entity_end < len(text):
+            # Adds the remainder of the text after the last end brace } "function(post_entity_text)"
+            modified_text.append(function(text[prev_entity_end : len(text)]))
+
+        return "".join(modified_text)
+
+    @staticmethod
+    def tokenize_around_mindmeld_annotations(text, function):
+        """Applied a function around the mindmeld annotation.
+
+        tokenize(pre_entity_text) + { + tokenize(entity_text) + |entity_name}
+            + tokenize(post_entity_text)
+
+        Args:
+            text (str): Original sentence with markup to modify.
+            function (function): Function to apply around the annotation
+        Returns:
+            tokens (List[dict]): List of tokens represented as dictionaries.
+        """
+        matches = TextPreparationPipeline.find_mindmeld_annotation_re_matches(text)
+
+        tokens = []
+        prev_entity_end = 0
+
+        for match in matches:
+            entity_start, entity_end = match.span()
+            entity_text = match.group(1)
+
+            # Adds tokens from text before the current entity and after the last entity
+            tokens_before_entity = function(text[prev_entity_end:entity_start])
+            TextPreparationPipeline.offset_token_start_values(
+                tokens=tokens_before_entity, offset=prev_entity_end
+            )
+            tokens.extend(tokens_before_entity)
+
+            # Adds tokens from text within the entity text
+            entity_text_start, _ = match.span(1)
+            entity_text = match.group(1)
+            tokens_within_entity_text = function(entity_text)
+            TextPreparationPipeline.offset_token_start_values(
+                tokens=tokens_within_entity_text, offset=entity_text_start
+            )
+            tokens.extend(tokens_within_entity_text)
+
+            # Update the previous entity ending index
+            prev_entity_end = entity_end
+
+        if prev_entity_end < len(text):
+            # Add tokens from the text after the last MindMeld entity
+            tokens_after_last_entity = function(text[prev_entity_end : len(text)])
+            TextPreparationPipeline.offset_token_start_values(
+                tokens=tokens_after_last_entity, offset=prev_entity_end
+            )
+            tokens.extend(tokens_after_last_entity)
+
+        tokens = TextPreparationPipeline.filter_out_space_text_tokens(tokens)
+        return tokens
+
+    @staticmethod
+    def offset_token_start_values(tokens: List[Dict], offset: int):
```

Comment:
Shouldn't this function be returning the tokens? Doesn't seem to be modifying it in-place?

Reply:
Double checked the function in a notebook, it does modify dictionary values in-place